<a href="https://colab.research.google.com/github/deepakahirwar/ADF_01/blob/main/Copy_of_Spotify_Case_Study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![sslogo](https://github.com/stratascratch/stratascratch.github.io/raw/master/assets/sslogo.jpg)

# Spotify Case Study

### Business Case

The Spotify marketing team is starting an advertising campaign. Their goal is to maximize the number of daily streams across all regions in the world. The only dataset they have is their worldwide daily song rankings which is a list of songs ranked by the number of user streams and organized by regions. As the manager of the marketing team, how should the team focus their advertising resources and what should they do to maximize the number of daily streams?

### Questions
#### A questions to answer to get you started:

- What percentage of total streams come from the top 10 artists? 
- What percentage of total streams come from the top 10 tracknames?
- Which artists had the most streams in 2017? List the top 10 artists and the total number of streams on Spotify.
- Do artists have several tracks in the top 100? 200? List the artists by the number of track names in the top 200 in descending order (you'll need to take into account the number of regions the artist is in).
- How do the number of streams in the top 10 differ than the number of streams in the top 50? top 100? top 200? Find the average number of streams in the top 10, 50, 100, 200.
- How many different artists are there in the top 100 vs top 101-200? Compare the number of artists in the top 100 vs the top 101-200.

#### A few open ended questions to guide your recommendations:

- Which artists should the marketing team invest in? Why? Feel free to add any additional analytical research. 
- What are your recommendations to maximize the number of streams? Would you focus on certain artists or a certain position? Expand your thoughts with some recommendations.

# New Section

### Imports

In [3]:

# import numpy, pandas, and spotify data(from files in canvas)
import pandas as pd
import csv 
import numpy as np
file = 'spotify_daily_rankings_2017_us.csv'
df = pd.read_csv(file)

#### Print the dataframe

In [4]:
df

position                               trackname            artist  \
0             1     Bad and Boujee (feat. Lil Uzi Vert)             Migos   
1             2                               Fake Love             Drake   
2             3                                 Starboy        The Weeknd   
3             4                                  Closer  The Chainsmokers   
4             5                           Black Beatles      Rae Sremmurd   
...         ...                                     ...               ...   
74195       196                                Portland             Drake   
74196       197                               All On Me      Devin Dawson   
74197       198  From The D To The A (feat. Lil Yachty)      Tee Grizzley   
74198       199                            Now Or Never            Halsey   
74199       200                So Good (& Metro Boomin)          Big Sean   

       streams                                                url        date  
0      1371493  https://open.spotify.com/track/4Km5HrUvYTaSUfi...  01-01-2017  
1      1180074  https://open.spotify.com/track/343YBumqHu19cGo...  01-01-2017  
2      1064351  https://open.spotify.com/track/5aAx2yezTd8zXrk...  01-01-2017  
3      1010492  https://open.spotify.com/track/7BKLCZ1jbUBVqRi...  01-01-2017  
4       874289  https://open.spotify.com/track/6fujklziTHa8uoM...  01-01-2017  
...        ...                                                ...         ...  
74195   184537  https://open.spotify.com/track/2bjwRfXMk4uRgOD...  01-09-2018  
74196   184519  https://open.spotify.com/track/2mfahQ0EaaZWq4c...  01-09-2018  
74197   184428  https://open.spotify.com/track/2NyrXRn4tancYPW...  01-09-2018  
74198   184394  https://open.spotify.com/track/7i2DJ88J7jQ8K7z...  01-09-2018  
74199   184149  https://open.spotify.com/track/0fWBFDRTIuk8ZgN...  01-09-2018  

[74200 rows x 6 columns]

## #### View the first 5 rows of the dataframe

In [80]:
print(df.iloc[0])


position                                                   196
trackname                                        Dennis Rodman
artist                                                mansionz
streams                                                 166943
url          https://open.spotify.com/track/73cAKC1NbxHuFPc...
date                                       2017-03-17 00:00:00
Name: 15195, dtype: object


####Convert date column to be type datetime

In [84]:
df['date'] = df['date'].apply(pd.to_datetime)
df.info()

df['date'] = pd.to_datetime(df['date'])


<class 'pandas.core.frame.DataFrame'>
Int64Index: 886 entries, 15195 to 39415
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   position   886 non-null    int64         
 1   trackname  884 non-null    object        
 2   artist     884 non-null    object        
 3   streams    886 non-null    int64         
 4   url        886 non-null    object        
 5   date       886 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 48.5+ KB


#### What percentage of total streams come from the top 10 artists?

We perform a groupby for an artist and then sum all daily stream counts to obtain a total stream count.
After that we sort by values. 

Notice here that we have a Series so sort_values does not need a name of a column to work.

Most answers will have the following syntax:



```
dataframe.groupby("column_name")["column_name"]\
         .aggragation_method()\ # i.e. sum(), mean(), etc.
         .sort_values(ascending=False) 
```



In [89]:
top_artists = df.groupby("artist")["streams"].sum()\
                       .sort_values(ascending=False)
print(top_artists)   
superstar_stream_count = top_artists[:10].sum()
total_stream_count     = top_artists.sum()
print(superstar_stream_count / total_stream_count)


artist
DJ Khaled         1971444
Mariah Carey      1722225
Future            1715855
French Montana    1530548
Kodak Black       1479849
                   ...   
DeJ Loaf           152420
Bobby McFerrin     151217
August Alsina      149475
Sam Cooke          145484
Phil Collins       143813
Name: streams, Length: 487, dtype: int64
0.06600220629599009


The percentage of streams come from the top 10 artist is following: 
**6.60%**


#### What percentage of total streams come from the top 10 tracknames?

Same logic as above except we group by trackname instead of artist.

In [90]:

top_tracks = df.groupby("trackname")["streams"].sum().sort_values(ascending=False)
    
top_track_stream_count = top_tracks[:10].sum()

total_stream_count     = top_tracks.sum()

print(top_track_stream_count / total_stream_count)

0.06744051250006494


The percentage of streams come from the top 10 artist is following: 
**6.67%**


#### Which artists had the most streams in 2017? List the top 10 artists and the total number of streams on Spotify


Please note that to split python code into multiple lines we sometimes need to use the \ symbol.

In [96]:
```sql
SELECT
    artist,
    SUM(streams) AS total_streams
FROM datasets.spotify_daily_rankings_2017_us
WHERE DATE_PART('year', date :: DATE) = 2017
GROUP BY artist
ORDER BY total_streams DESC
LIMIT 10

SyntaxError: ignored

#### Do artists have several tracks in the top 100? 200? List the artists by the number of track names in the top 200 in descending order.

The nunique method is the version of the count method to use

In [100]:
 df.groupby("artist")["trackname"]\
          .nunique()\
          .sort_values(ascending=False)

artist
Lil Peep           2
Noah Cyrus         2
DJ Snake           2
Tee Grizzley       2
Taylor Swift       2
                  ..
Imagine Dragons    1
ILoveMakonnen      1
Hozier             1
Hearts & Colors    1
mansionz           1
Name: trackname, Length: 487, dtype: int64

#### How do the number of streams in the top 10 differ than the number of streams in the top 50? top 100? top 200? Find the average number of streams in the top 10, 50, 100, 200.

To solve this task we used a python loop.

We iterate over a set of 4 limits and calculate the average for each position limit.

We calculate the average with a filter `dataframe.position <= limit`.

In [101]:
for limit in [10, 50, 100, 200]:
    ave_streams = df[df.position <= limit].streams.mean()
    print("For position limit {0} we got {1} average streams".format(limit, ave_streams))

For position limit 10 we got 1007328.5 average streams
For position limit 50 we got 628136.925 average streams
For position limit 100 we got 408480.2710622711 average streams
For position limit 200 we got 262689.38826185104 average streams


#### How many different artists are there in the top 100 vs top 101-200? Compare the number of artists in the top 100 vs the top 101-200.

Remember to use the nunique() method

In [99]:
print(df[:100])

       position                                          trackname  \
15195       196                                      Dennis Rodman   
2287         88          i hate u, i love u (feat. olivia o'brien)   
678          79          i hate u, i love u (feat. olivia o'brien)   
113         114  We Are Young (feat. Janelle Mon‰ö_å¡e) - feat....   
35479        80                                           do re mi   
...         ...                                                ...   
2563        164                                          On My Way   
26399       200                        Raincoat (feat. Shy Martin)   
70537       138  You're A Mean One, Mr. Grinch - From Dr. Seuss...   
66753       154  You're A Mean One, Mr. Grinch - From Dr. Seuss...   
32993       194                                        Craving You   

                  artist  streams  \
15195           mansionz   166943   
2287               gnash   245870   
678                gnash   242308   
113        

#### Which artists should the marketing team invest in? Why? Feel free to add any additional analytical research.


In [119]:

statistics = []
for artist, artist_row_indices in groups.items():
    count = artist_row_indices.values.shape[0]
    if count <= 1:
        continue
         streams = df.iloc[artist_row_indices]["streams"].values
          x = np.arange(count)
  A = np.vstack([x, np.ones(x.shape[0])]).T
   y = streams
    m, c = np.linalg.lstsq(A, y, rcond=None)[0]
       score = m * count
     statistics_dataframe = pd.df(statistics, 
                                    columns=["artist", "count", "slope", "offset", "score"])
statistics_dataframe.head()

statistics_dataframe\
   .query("count >= 10 and count <= 100")\
   .sort_values(by=['score'], ascending=False)\
   .iloc[:50]

IndentationError: ignored